src/main/java/io/cdap/plugin/gcp/bigtable/source/BigtableSource.java

```diff
@@ -0,0 +1,249 @@
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    outputSchema = context.getOutputSchema();
+  }
+
+  /**
+   * Converts <code>JsonObject</code> to <code>StructuredRecord</code> for every record
+   * retrieved from the Bigtable.
+   *
+   * @param input   input record
+   * @param emitter emitting the transformed record into downstream nodes.
+   */
+  @Override
+  public void transform(KeyValue<ImmutableBytesWritable, Result> input, Emitter<StructuredRecord> emitter) {
+    try {
+      StructuredRecord record = convertInputToRecord(input);
+      emitter.emit(record);
+    } catch (Exception e) {
+      switch (config.getErrorHandling()) {
+        case SKIP:
+          LOG.warn("Failed to process message, skipping it", e);
+          break;
+        case FAIL_PIPELINE:
+          throw new RuntimeException("Failed to process message", e);
+        default:
+          throw new IllegalStateException(String.format("Unknown error handling strategy '%s'",
+                                                            config.getErrorHandling()));
+      }
+    }
+  }
+
+  private StructuredRecord convertInputToRecord(KeyValue<ImmutableBytesWritable, Result> input) {
+    Result result = input.getValue();
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(outputSchema);
+    List<Schema.Field> fields = Objects.requireNonNull(outputSchema.getFields());
+    for (Schema.Field field : fields) {
+      String fieldName = field.getName();
+      Schema fieldSchema = field.getSchema();
+      if (fieldName.equals(config.keyAlias)) {
+        Object value = convertBytesToField(input.getKey().get(), fieldSchema);
+        recordBuilder.set(fieldName, value);
+      } else {
+        HBaseColumn hBaseColumn = HBaseColumn.fromFullName(fieldName);
+        byte[] valueBytes =
+          result.getValue(Bytes.toBytes(hBaseColumn.getFamily()), Bytes.toBytes(hBaseColumn.getQualifier()));
+        Object value = convertBytesToField(valueBytes, fieldSchema);
+        recordBuilder.set(fieldName, value);
+      }
+    }
+    return recordBuilder.build();
+  }
+
+  private Object convertBytesToField(byte[] bytes, Schema fieldSchema) {
+    if (fieldSchema.isNullable()) {
+      if (null == bytes) {
+        return null;
+      }
+      return convertBytesToField(bytes, fieldSchema.getNonNullable());
+    }
+    switch (fieldSchema.getType()) {
+      case NULL:
+        return null;
+      case STRING:
+        return Bytes.toString(bytes);
+      case BYTES:
+        return bytes;
+      case INT:
+        return Bytes.toInt(bytes);
+      case LONG:
+        return Bytes.toLong(bytes);
+      case FLOAT:
+        return Bytes.toFloat(bytes);
+      case DOUBLE:
+        return Bytes.toDouble(bytes);
+      case BOOLEAN:
+        return Bytes.toBoolean(bytes);
+      default:
+        throw new UnexpectedFormatException("Field type '" + fieldSchema.getType() + "' is not supported");
```

Comment:
message should include which field has the problem, otherwise the user won't know what to do.

Field '%s' is of unsupported type '%s'.

Reply:
fixed